File: backend/app/schemas/community.py
from datetime import date, datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field
from app.database.models.community import (
    CommunityJoinType,
    CommunityProfileType,
    CommunityUrgency,
)
from app.database.models.community_member import CommunityMemberRole

# Paleta cerrada y elegante para la portada automática (fondo de color
# + avatares de los miembros encima, ver frontend CommunityCover.tsx).
# Deliberadamente pequeña: nada de un selector de color libre, para
# mantener el aspecto premium/curado en todas las comunidades.
CommunityCoverColor = Literal[
    "sage", "cream", "stone", "sand", "smoke", "forest"
]


class CommunityPreferencesCreate(BaseModel):
    cleanliness: str = Field(min_length=1, max_length=150)
    atmosphere: str = Field(min_length=1, max_length=150)
    visits: str = Field(min_length=1, max_length=150)
    sleepovers: str = Field(min_length=1, max_length=150)
    smoking: str = Field(min_length=1, max_length=150)
    pets: str = Field(min_length=1, max_length=150)
    rules: str = Field(min_length=1, max_length=150)
    lifestyle: str = Field(min_length=1, max_length=150)


class CommunityPreferencesUpdate(BaseModel):
    cleanliness: str | None = Field(
        default=None,
        min_length=1,
        max_length=150,
    )
    atmosphere: str | None = Field(
        default=None,
        min_length=1,
        max_length=150,
    )
    visits: str | None = Field(
        default=None,
        min_length=1,
        max_length=150,
    )
    sleepovers: str | None = Field(
        default=None,
        min_length=1,
        max_length=150,
    )
    smoking: str | None = Field(
        default=None,
        min_length=1,
        max_length=150,
    )
    pets: str | None = Field(
        default=None,
        min_length=1,
        max_length=150,
    )
    rules: str | None = Field(
        default=None,
        min_length=1,
        max_length=150,
    )
    lifestyle: str | None = Field(
        default=None,
        min_length=1,
        max_length=150,
    )


class CommunityPreferencesResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    community_id: int

    cleanliness: str
    atmosphere: str
    visits: str
    sleepovers: str
    smoking: str
    pets: str
    rules: str
    lifestyle: str

    created_at: datetime
    updated_at: datetime

class CommunityMemberUserResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    first_name: str
    last_name: str
    avatar_url: str | None = None


class CommunityMemberResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    user_id: UUID
    role: CommunityMemberRole
    joined_at: datetime
    user: CommunityMemberUserResponse


class CommunityCreate(BaseModel):
    name: str = Field(min_length=3, max_length=120)
    description: str = Field(min_length=20)
    city: str = Field(min_length=1, max_length=100)
    province: str | None = Field(default=None, max_length=100)
    neighborhood: str | None = Field(default=None, max_length=120)
    max_members: int = Field(ge=2, le=12)

    preferences: CommunityPreferencesCreate
    join_type: CommunityJoinType = CommunityJoinType.REQUEST

    open_spots: int = Field(default=1, ge=0, le=12)

    urgency: CommunityUrgency = CommunityUrgency.NORMAL

    profile_type: CommunityProfileType
    profile_description: str | None = Field(
        default=None,
        max_length=500,
    )

    monthly_rent: int | None = Field(
        default=None,
        ge=0,
        le=10000,
    )

    deposit: int | None = Field(
        default=None,
        ge=0,
        le=30000,
    )

    move_in_date: date | None = None

    room_description: str | None = Field(
        default=None,
        max_length=2000,
    )

    cover_color: CommunityCoverColor = "sage"


class CommunityUpdate(BaseModel):
    name: str | None = Field(
        default=None,
        min_length=3,
        max_length=120,
    )
    description: str | None = Field(
        default=None,
        min_length=20,
    )
    city: str | None = Field(
        default=None,
        min_length=1,
        max_length=100,
    )
    province: str | None = Field(
        default=None,
        max_length=100,
    )
    neighborhood: str | None = Field(
        default=None,
        max_length=120,
    )
    max_members: int | None = Field(
        default=None,
        ge=2,
        le=12,
    )

    preferences: CommunityPreferencesUpdate | None = None
    join_type: CommunityJoinType | None = None

    open_spots: int | None = Field(
        default=None,
        ge=0,
        le=12,
    )

    urgency: CommunityUrgency | None = None

    profile_type: CommunityProfileType | None = None
    profile_description: str | None = Field(
        default=None,
        max_length=500,
    )

    monthly_rent: int | None = Field(
        default=None,
        ge=0,
        le=10000,
    )

    deposit: int | None = Field(
        default=None,
        ge=0,
        le=30000,
    )

    move_in_date: date | None = None

    room_description: str | None = Field(
        default=None,
        max_length=2000,
    )

    cover_color: CommunityCoverColor | None = None


class CommunityResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str
    description: str
    city: str
    province: str | None = None
    neighborhood: str | None = None
    max_members: int

    join_type: CommunityJoinType

    open_spots: int
    urgency: CommunityUrgency
    profile_type: CommunityProfileType
    profile_description: str | None = None
    monthly_rent: int | None = None
    total_monthly_rent: int | None = None
    deposit: int | None = None
    move_in_date: date | None = None
    room_description: str | None = None

    cover_color: str
    cover_image_url: str | None = None

    owner_id: UUID
    is_active: bool
    created_at: datetime
    updated_at: datetime

    preferences: CommunityPreferencesResponse | None = None

    member_count: int = 0
    is_member: bool = False
    current_user_role: CommunityMemberRole | None = None
    is_full: bool = False

    members: list[CommunityMemberResponse] = Field(
        default_factory=list
    )

File: frontend/src/types/community.ts
import type { CompatibilityScore } from "./compatibilityScore";

export interface CommunityPreferences {
  id: number;
  community_id: number;
  cleanliness: string;
  atmosphere: string;
  visits: string;
  sleepovers: string;
  smoking: string;
  pets: string;
  rules: string;
  lifestyle: string;
  created_at: string;
  updated_at: string;
}

export interface CommunityPreferencesCreate {
  cleanliness: string;
  atmosphere: string;
  visits: string;
  sleepovers: string;
  smoking: string;
  pets: string;
  rules: string;
  lifestyle: string;
}

export interface CommunityPreferencesUpdate {
  cleanliness?: string;
  atmosphere?: string;
  visits?: string;
  sleepovers?: string;
  smoking?: string;
  pets?: string;
  rules?: string;
  lifestyle?: string;
}

export type CommunityCoverColor =
  | "sage"
  | "cream"
  | "stone"
  | "sand"
  | "smoke"
  | "forest";

export interface Community {
  id: number;
  name: string;
  description: string;
  city: string;
  province: string | null;
  neighborhood: string | null;
  max_members: number;
  owner_id: string;
  is_active: boolean;
  created_at: string;
  updated_at: string;
  preferences: CommunityPreferences | null;
  join_type: CommunityJoinType;
  open_spots: number;
  urgency: CommunityUrgency;
  profile_type: CommunityProfileType;
  profile_description: string | null;
  monthly_rent: number | null;
  total_monthly_rent: number | null;
  deposit: number | null;
  move_in_date: string | null;
  room_description: string | null;
  cover_color: CommunityCoverColor;
  cover_image_url: string | null;
  member_count: number;
  is_member: boolean;
  current_user_role: CommunityMemberRole | null;
  is_full: boolean;
  is_saved: boolean;
  members: CommunityMember[];
  average_compatibility: CompatibilityScore | null;
}

export interface CommunityCreate {
  name: string;
  description: string;
  city: string;
  province?: string | null;
  neighborhood?: string | null;
  max_members: number;
  preferences: CommunityPreferencesCreate;
  join_type: CommunityJoinType;
  open_spots: number;
  urgency: CommunityUrgency;
  profile_type: CommunityProfileType;
  profile_description?: string | null;
  monthly_rent?: number | null;
  deposit?: number | null;
  move_in_date?: string | null;
  room_description?: string | null;
  cover_color?: CommunityCoverColor;
}

export interface CommunityUpdate {
  name?: string;
  description?: string;
  city?: string;
  province?: string | null;
  neighborhood?: string | null;
  max_members?: number;
  preferences?: CommunityPreferencesUpdate;
  join_type?: CommunityJoinType;
  open_spots?: number;
  urgency?: CommunityUrgency;
  profile_type?: CommunityProfileType;
  profile_description?: string | null;
  monthly_rent?: number | null;
  deposit?: number | null;
  move_in_date?: string | null;
  room_description?: string | null;
  cover_color?: CommunityCoverColor;
}

export type CommunityProfileType =
  | "STUDENTS"
  | "YOUNG_PROFESSIONALS"
  | "DIGITAL_NOMADS"
  | "WORKERS"
  | "EXAM_CANDIDATES"
  | "INTERNATIONAL_STUDENTS"
  | "MIXED"
  | "OTHER";

export interface GetCommunitiesParams {
  city?: string;
  province?: string;
  profile_type?: CommunityProfileType;
  skip?: number;
  limit?: number;
}

export interface CommunityMemberUser {
  id: string;
  first_name: string;
  last_name: string;
  avatar_url: string | null;
  age: number | null;
  is_email_verified: boolean;
}

export interface CommunityMember {
  id: number;
  user_id: string;
  role: CommunityMemberRole;
  joined_at: string;
  user: CommunityMemberUser;
}

export type CommunityJoinType = "OPEN" | "REQUEST";
export type CommunityMemberRole = "OWNER" | "MEMBER";
export type CommunityUrgency = "NORMAL" | "SOON" | "URGENT";

export interface CommunityMessageSender {
  id: string;
  first_name: string;
  last_name: string;
  avatar_url: string | null;
}

export interface CommunityMessageReplyPreview {
  id: number;
  content: string;
  sender_id: string;
  sender_first_name: string;
}

export interface CommunityMessage {
  id: number;
  community_id: number;
  content: string;
  created_at: string;
  updated_at: string;
  sender: CommunityMessageSender;
  reply_to: CommunityMessageReplyPreview | null;
  like_count: number;
  liked_by_me: boolean;
}

export interface CommunityMessageCreate {
  content: string;
  reply_to_id?: number | null;
}

export interface GetCommunityMessagesParams {
  skip?: number;
  limit?: number;
}

export interface CommunityMemberContribution {
  member_id: number;
  user_id: string;
  first_name: string;
  last_name: string;
  role: CommunityMemberRole;
  monthly_contribution: number | null;
  contribution_percentage: number | null;
}

export interface CommunityRentSplit {
  total_monthly_rent: number | null;
  total_configured: number;
  remaining_amount: number | null;
  contributions: CommunityMemberContribution[];
}

export interface CommunityRentContributionUpdate {
  member_id: number;
  monthly_contribution: number | null;
}

export interface CommunityRentSplitUpdate {
  total_monthly_rent: number | null;
  contributions: CommunityRentContributionUpdate[];
}